Enable Proguard for android

hristoterezov opened the pull request.

paweldomas commented:
How did you come up with the specific options ? is the proguard config file based on something ?

hristoterezov commented:
@paweldomas we had some options even before I used this as base and added more rules there. About the new rules - I was just testing and at some point you see error (... class is missing) then I added the rules to not remove this specific classes because of the errors.

paweldomas commented on android/app/build.gradle:
```diff
@@ -27,8 +27,12 @@ android {
 
     buildTypes {
+        debug {
+            minifyEnabled true
```
I'm a bit worried that obfuscating the code may cause problems with crashlytics. Can you check this guide ?
https://docs.fabric.io/android/crashlytics/dex-and-proguard.html

paweldomas commented on android/app/proguard-rules.pro:
```diff
@@ -18,7 +18,7 @@
 # Disabling obfuscation is useful if you collect stack traces from production crashes
 # (unless you are using a system that supports de-obfuscate the stack traces).
--dontobfuscate
+# -dontobfuscate
```
I think we don't want to obfuscate release builds (see my comment above). Maybe it's fine to obfuscate when the crashlytics guide linked above is followed

hristoterezov commented:
@paweldomas I've added the rules.

lyubomir commented on android/app/proguard-rules-debug.pro:
```diff
@@ -0,0 +1,93 @@
+# Add project specific ProGuard rules here.
```
There's content in this file which looks to me to be taken from somewhere. I imagine the origin is proguard-rules.pro. Seeing that `proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'` lists a couple of files, I imagine we can have a common ProGuard rule file for debug and release builds. Alternatively, I think I read ProGuard has an `-include` so we could include the common file in the debug and release-specific files.

hristoterezov replied:
Done!

lyubomir commented:
APK file sizes before this PR:
Debug: 18930053 (18M)
Release 18077977 (17M)

After:
Debug: 17974712 (17M)
Release: 17057405 (16M)

hristoterezov commented on android/app/proguard-rules.pro:
```diff
@@ -48,6 +48,15 @@
 -keepclassmembers class *  { @com.facebook.react.uimanager.annotations.ReactProp <methods>; }
 -keepclassmembers class *  { @com.facebook.react.uimanager.annotations.ReactPropGroup <methods>; }
 
+-keep class com.facebook.react.bridge.ModuleRegistryHolder { *; }
```
Note that before this PR proguard wasn't enabled for any of the builds. So I guess it will work before this PR even if you delete all the rules. At least this is my understanding, please correct me if I'm wrong. This PR adds `minifyEnabled true` which enables proguard and then you start seeing the errors.

I don't have solid reasoning behind every rule that I have added. As I mentioned in previous comments I started testing to discover the classes that are missing after enabling minify and I added exception for every error that I have discovered.

hristoterezov commented on android/app/proguard-rules.pro:
```diff
@@ -74,3 +84,10 @@
 
+-keep class org.webrtc.** { *; }
+
+
```
what do you mean by "what is being ignored here..."? `class org.webrtc.** { *; }` is ignored, isn't it? I'm having a hard time to understand your comment. BTW @lyubomir already added a comment for the whole block of rules added after we changed the value of `minifyEnabled`. Let me know if you want me to add additional comments.

hristoterezov commented:
@saghul 
> Agreed, this change doesn't seem related?

Could you please explain?

saghul commented:
> Could you please explain?

Before these changes the `dontobfuscate` thing was always set for release builds. No that's no longer the case. Is this related to any of the other changes in the PR? I fail to see how, can you clarify? I worry our Crashlytics traces are going to be useless otherwise. Thanks!

hristoterezov commented:
I, @lyubomir and @paweldomas discussed this offline. We agreed that we are going to try to enable what we can from the Proguard including obfuscation. Crashlytics should be fine according to:
https://docs.fabric.io/android/crashlytics/dex-and-proguard.html
